File: src/queries/wtlive.ts
import {
  keepPreviousData,
  useInfiniteQuery,
  useMutation,
  useQuery,
  useQueryClient,
  type InfiniteData,
  type QueryKey,
} from '@tanstack/react-query';
import { useEffect } from 'react';
import { listenEvent } from '@/lib/events';
import { call, hasBackend, toAppError } from '@/lib/tauri';
import { useUi } from '@/store/ui';
import {
  EVENTS,
  type AppError,
  type ConflictPolicy,
  type FollowEntry,
  type FollowKind,
  type HangarSkin,
  type InstallMode,
  type InstallStarted,
  type NetStatus,
  type SearchParams,
  type SearchResult,
  type WtLiveSkin,
} from '@/types';
import { HANGAR_KEY } from './hangar';

export const WTLIVE_KEY = ['wtlive'] as const;
export const FOLLOWING_KEY = ['following'] as const;

/** Errors that mean "WT Live can't be reached from here" (drives the offline state). */
const OFFLINE_CODES = new Set(['network', 'unsupported', 'noBackend']);

/** Runs a WT Live call and keeps `ui.online` in step with how it went. */
async function tracked<T>(run: () => Promise<T>): Promise<T> {
  if (!hasBackend()) {
    useUi.getState().setOnline(false);
    throw { code: 'noBackend', message: 'WT Live needs the desktop app' } satisfies AppError;
  }
  try {
    const result = await run();
    useUi.getState().setOnline(true);
    return result;
  } catch (e) {
    const error = toAppError(e);
    if (OFFLINE_CODES.has(error.code)) useUi.getState().setOnline(false);
    throw error;
  }
}

export const isOfflineError = (e: AppError | null | undefined) => !!e && OFFLINE_CODES.has(e.code);

/** Keeps `ui.online` in step with the backend's `net://status` events; mounted once in App. */
export function useNetStatusEvents() {
  useEffect(() => {
    let disposed = false;
    let unlisten: (() => void) | undefined;
    void listenEvent<NetStatus>(EVENTS.netStatus, (status) => useUi.getState().setOnline(status.online))
      .then((fn) => {
        if (disposed) fn();
        else unlisten = fn;
      })
      .catch((e: unknown) => console.error('[livery] net://status listener failed', e));
    return () => {
      disposed = true;
      unlisten?.();
    };
  }, []);
}

/** Explore grid. Filters combine (AND); previous results stay on screen while the next page loads. */
export function useWtLiveSearch(params: SearchParams) {
  return useQuery<SearchResult, AppError>({
    queryKey: [...WTLIVE_KEY, 'search', params],
    queryFn: () => tracked(() => call<SearchResult>('wtlive_search', { params })),
    placeholderData: keepPreviousData,
    staleTime: 5 * 60_000,
  });
}

/**
 * Explore grid, page after page (BUILD_PLAN open decision 2): the backend fetches, caches and
 * filters WT Live locally and answers pages of results; call `fetchNextPage()` when the virtual
 * grid nears the end. `total` and `tookMs` come from the first page.
 */
export function useWtLiveSearchPages(params: Omit<SearchParams, 'page'>) {
  return useInfiniteQuery<SearchResult, AppError, InfiniteData<SearchResult, number>, QueryKey, number>({
    queryKey: [...WTLIVE_KEY, 'search-pages', params],
    queryFn: ({ pageParam }) => tracked(() => call<SearchResult>('wtlive_search', { params: { ...params, page: pageParam } })),
    initialPageParam: 0,
    getNextPageParam: (last, pages) => {
      const loaded = pages.reduce((n, p) => n + p.items.length, 0);
      return last.items.length > 0 && loaded < last.total ? pages.length : undefined;
    },
    placeholderData: keepPreviousData,
    staleTime: 5 * 60_000,
  });
}

/** Full post (images, files) for the Skin detail. */
export function useWtLivePost(id: string | null) {
  return useQuery<WtLiveSkin, AppError>({
    queryKey: [...WTLIVE_KEY, 'post', id],
    queryFn: () => tracked(() => call<WtLiveSkin>('wtlive_post', { id })),
    enabled: id !== null,
    staleTime: 5 * 60_000,
  });
}

/** New skins from followed vehicles/authors since they were last seen. */
export function useFollowingNew(entries: FollowEntry[] | undefined) {
  const vehicles = (entries ?? []).filter((e) => e.kind === 'vehicle').map((e) => e.id);
  const authors = (entries ?? []).filter((e) => e.kind === 'author').map((e) => e.id);
  return useQuery<WtLiveSkin[], AppError>({
    queryKey: [...WTLIVE_KEY, 'following-new', vehicles, authors],
    queryFn: () => tracked(() => call<WtLiveSkin[]>('wtlive_following_new', { vehicles, authors })),
    enabled: entries !== undefined && (vehicles.length > 0 || authors.length > 0),
    staleTime: 5 * 60_000,
  });
}

/** Followed vehicles and authors (local data; works offline). */
export function useFollowing() {
  return useQuery<FollowEntry[], AppError>({
    queryKey: FOLLOWING_KEY,
    queryFn: () => (hasBackend() ? call<FollowEntry[]>('following_list') : Promise.resolve([])),
  });
}

export interface SetFollowArgs {
  kind: FollowKind;
  id: string;
  name: string;
  follow: boolean;
  /**
   * RFC 3339. With `follow`, the entry gets this `lastSeenAt` (new or already followed): the Undo
   * of an unfollow passes the entry's old value so its "N new" comes back exactly.
   */
  lastSeenAt?: string;
}

/** Follows or unfollows a vehicle or an author (`following_set`); resolves with the whole list. */
export function useSetFollow() {
  const qc = useQueryClient();
  return useMutation<FollowEntry[], AppError, SetFollowArgs>({
    mutationFn: ({ lastSeenAt, ...args }) =>
      call<FollowEntry[]>('following_set', lastSeenAt === undefined ? { ...args } : { ...args, lastSeenAt }),
    onSuccess: (list) => qc.setQueryData(FOLLOWING_KEY, list),
  });
}

/**
 * Leaving the Following tab resets its "N new" count. `following-new` holds only the asked ids
 * (not their `lastSeenAt`), so it is invalidated too: its cached skins are no longer new.
 */
export function useMarkFollowingSeen() {
  const qc = useQueryClient();
  return useMutation<FollowEntry[], AppError, void>({
    mutationFn: () => call<FollowEntry[]>('following_mark_seen'),
    onSuccess: (list) => {
      qc.setQueryData(FOLLOWING_KEY, list);
      return qc.invalidateQueries({ queryKey: [...WTLIVE_KEY, 'following-new'] });
    },
  });
}

/** Starts a download + install; progress arrives as `install://progress` (see src/store/installs.ts). */
export function installFromWtLive(skinId: string, mode: InstallMode, conflict?: ConflictPolicy): Promise<InstallStarted> {
  return tracked(() => call<InstallStarted>('install_from_wtlive', conflict ? { skinId, mode, conflict } : { skinId, mode }));
}

/**
 * Try in game → Keep (becomes a normal hangar skin) or Discard (removed, game files restored).
 * `skinId` is the **WT Live** skin id (the hangar skin's `sourceId`), not a hangar id.
 */
export function useFinalizeTry() {
  const qc = useQueryClient();
  return useMutation<HangarSkin | null, AppError, { skinId: string; keep: boolean }>({
    mutationFn: (args) => call<HangarSkin | null>('finalize_try', args),
    onSettled: () => qc.invalidateQueries({ queryKey: HANGAR_KEY }),
  });
}

File: src-tauri/src/wtlive.rs
//! WT Live (M5): listing/post parsing behind a swappable trait, 24 h cache, 1 req/s rate limit with a
//! Livery user-agent; failures surface as `net://status` offline, never a crash.
//!
//! Stubs until an HTTP client (and an HTML parser) are approved: every network command answers
//! `unsupported` and reports WT Live as offline. Following is local data and works.

use crate::error::{AppError, AppResult, ErrorCode};
use crate::model::{
    ConflictPolicy, FollowEntry, FollowKind, HangarSkin, InstallMode, InstallStarted, SearchParams, SearchResult,
    WtLiveSkin,
};
use tauri::AppHandle;

pub const NET_STATUS_EVENT: &str = "net://status";

fn not_yet<T>(what: &str) -> AppResult<T> {
    Err(AppError::new(ErrorCode::Internal, format!("{what} is not implemented yet")))
}

#[tauri::command]
pub async fn wtlive_search(app: AppHandle, params: SearchParams) -> AppResult<SearchResult> {
    let _ = (app, params);
    not_yet("wtlive_search")
}

#[tauri::command]
pub async fn wtlive_post(app: AppHandle, id: String) -> AppResult<WtLiveSkin> {
    let _ = (app, id);
    not_yet("wtlive_post")
}

#[tauri::command]
pub async fn wtlive_following_new(
    app: AppHandle,
    vehicles: Vec<String>,
    authors: Vec<String>,
) -> AppResult<Vec<WtLiveSkin>> {
    let _ = (app, vehicles, authors);
    not_yet("wtlive_following_new")
}

#[tauri::command]
pub async fn install_from_wtlive(
    app: AppHandle,
    skin_id: String,
    mode: InstallMode,
    conflict: Option<ConflictPolicy>,
) -> AppResult<InstallStarted> {
    let _ = (app, skin_id, mode, conflict);
    not_yet("install_from_wtlive")
}

/// Try in game: keep → the temporary install becomes a normal hangar skin; discard → removed.
#[tauri::command]
pub async fn finalize_try(app: AppHandle, skin_id: String, keep: bool) -> AppResult<Option<HangarSkin>> {
    let _ = (app, skin_id, keep);
    not_yet("finalize_try")
}

#[tauri::command]
pub async fn following_list(app: AppHandle) -> AppResult<Vec<FollowEntry>> {
    let _ = app;
    not_yet("following_list")
}

/// Follows (`follow: true`) or unfollows a vehicle or an author; returns the whole list.
#[tauri::command]
pub async fn following_set(
    app: AppHandle,
    kind: FollowKind,
    id: String,
    name: String,
    follow: bool,
) -> AppResult<Vec<FollowEntry>> {
    let _ = (app, kind, id, name, follow);
    not_yet("following_set")
}

/// Marks every followed entry as seen now (the Following tab's "N new" resets).
#[tauri::command]
pub async fn following_mark_seen(app: AppHandle) -> AppResult<Vec<FollowEntry>> {
    let _ = app;
    not_yet("following_mark_seen")
}
